src/rdkafka_sasl_oauthbearer_oidc.c

```diff
@@ -0,0 +1,397 @@
+ *          If succeed, return rd_true with NULL in \p *error,
+ *          If failed, return rd_false with error message in \p *error.
+ *
+ * @locality Any thread.
+ */
+static rd_bool_t build_headers (rd_kafka_conf_t *conf,
+                                struct curl_slist **headers,
+                                char **error) {
+        static const char *miss_client_id_error_message = "client_id is "
+                "a required field";
+        static const char *miss_client_secret_error_message = "client_secret "
+                "is a required field";
+        static const char *accept_header = "Accept:application/json";
+        const char *client_id;
+        const char *client_secret;
+        char *authorization_base64_header;
+
+        client_id = conf->sasl.oauthbearer.client_id;
+
+        if (client_id == NULL) {
+                *error = rd_malloc(strlen(miss_client_id_error_message)
+                                   + 1);
+                strcpy(*error, miss_client_id_error_message);
+                return rd_false;
+        }
+
+        client_secret = conf->sasl.oauthbearer.client_secret;
+        if (client_secret == NULL) {
+                *error = rd_malloc(strlen(miss_client_secret_error_message)
+                                   + 1);
+                strcpy(*error, miss_client_secret_error_message);
+                return rd_false;
+        }
+
+        authorization_base64_header =
+                build_authorization_base64_header(client_id,
+                                                  client_secret,
+                                                  &*error);
+        if (authorization_base64_header == NULL)
+                return rd_false;
+
+        *headers = curl_slist_append(*headers, accept_header);
+        *headers = curl_slist_append(*headers, authorization_base64_header);
+        rd_free(authorization_base64_header);
+        return rd_true;
+}
+
+
+/**
+ * @brief Implementation of Oauth/OIDC token refresh call back function,
+ *        will receive the json response after HTTP call from token provider,
+ *        then extract the jwt from the json response, and forward it to
+ *        the broker.
+ */
+void rd_kafka_conf_set_oauthbearer_oidc_token_refresh_cb (rd_kafka_t *rk,
+        const char *oauthbearer_config, void *opaque) {
+        static const char *data = "grant_type=client_credentials,scope=%s";
+        static const char *token_key = "access_token";
+        static const long timeout = 20;
+        static const size_t retry = 3;
+
+        cJSON *json = NULL;
+        rd_http_error_t *herr;
+        char *jwt_token = NULL;
+        struct curl_slist *headers = NULL;
+        char *data_to_token;
+        size_t data_to_token_size;
+        const char *scope;
+        const char *token_url;
+        rd_bool_t build_headers_status;
+        char *error;
+
+        token_url = rk->rk_conf.sasl.oauthbearer.token_endpoint_url;
```

Comment:
We shouldnt even get here if this is not set, should be checked in conf_finalize